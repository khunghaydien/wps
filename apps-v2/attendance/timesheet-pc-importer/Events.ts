import uuid from 'uuid';

import Employee from '@widgets/dialogs/ProxyEmployeeSelectDialog/models/Employee';

import { create } from '@attendance/libraries/Event';

export default {
  selectedEmployee: create<Employee>(`SELECTED_EMPLOYEE-${uuid()}`),
} as const;
