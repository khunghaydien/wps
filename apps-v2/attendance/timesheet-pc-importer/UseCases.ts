import UseCaseCollection from '@attendance/application/UseCaseCollection';
import { IUseCase as IImporterTimesheetSaveUseCase } from '@attendance/domain/combinedUseCases/importer/ICheckAndSaveTimesheetUseCase';
import { IUseCase as IDailyRecordFetchRestTimeReasonsUseCase } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';
import { IUseCase as IDailyRequestEarlyLeaveRequestFetchReasonsUseCase } from '@attendance/domain/useCases/dailyRequest/earlyLeaveRequest/IFetchReasonsUseCase';
import { IUseCase as IDailyRequestLateArrivalRequestFetchReasonsUseCase } from '@attendance/domain/useCases/dailyRequest/lateArrivalRequest/IFetchReasonsUseCase';
import { IUseCase as IGeneralFetchUserSettingUseCase } from '@attendance/domain/useCases/general/IFetchUserSettingUseCase';
import { IUseCase as IImporterTimesheetCheckUseCase } from '@attendance/domain/useCases/importer/timesheet/ICheckUseCase';
import { IUseCase as IFetchContractedWorkTimesUseCase } from '@attendance/domain/useCases/importer/timesheet/IFetchContractedWorkTimesUseCase';
import { IUseCase as ILeaveFetchListUseCase } from '@attendance/domain/useCases/leave/IFetchListUseCase';

export type UseCases = {
  fetchUserSetting: IGeneralFetchUserSettingUseCase;
  fetchRestTimeReasons: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchRestTimeReasonsForBulk: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchLeaves: ILeaveFetchListUseCase;
  fetchEarlyLeaveReasons: IDailyRequestEarlyLeaveRequestFetchReasonsUseCase;
  fetchLateArrivalReasons: IDailyRequestLateArrivalRequestFetchReasonsUseCase;
  fetchContractedWorkTimes: IFetchContractedWorkTimesUseCase;
  saveTimesheet: IImporterTimesheetSaveUseCase;
  checkTimesheet: IImporterTimesheetCheckUseCase;
};

const service = UseCaseCollection<UseCases>(
  "Timesheet-PC-importer's UseCase Service"
);

export default service;
