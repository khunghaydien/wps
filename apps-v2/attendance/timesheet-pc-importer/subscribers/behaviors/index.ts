import { Store } from 'redux';

import cleanTimesheetRecords from './cleanTimesheetRecords';
import fetchRestTimesBulk from './fetchRestTimesBulk';
import { bind } from '@attendance/libraries/Collection';

export const methods = {
  cleanTimesheetRecords,
  fetchRestTimesBulk,
};

export default (store: Store) => bind(methods, store);
