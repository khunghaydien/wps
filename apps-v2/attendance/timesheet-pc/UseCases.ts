import UseCaseCollection from '@attendance/application/UseCaseCollection';
import { IUseCase as ISubmitFixDailyRequestAndSaveDailyRecordUseCase } from '@attendance/domain/combinedUseCases/ISaveDailyRecordAndSubmitFixDailyRequestUseCase';
import { IUseCase as ISubmitFixDailyRequestWithClockOutUseCase } from '@attendance/domain/combinedUseCases/ISubmitFixDailyRequestWithClockOutUseCase';
import { IUseCase as IDailyRecordFetchRestTimeReasonsUseCase } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';
import { IUseCase as IDailyRecordSaveFieldsUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveFieldsUseCase';
import { IUseCase as IDailyRecordSaveUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveUseCase';
import { IUseCase as IFixDailyRequestCancelApprovalUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelApprovalUseCase';
import { IUseCase as IFixDailyRequestCancelSubmittedUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelSubmittedUseCase';
import { IUseCase as IFixDailyRequestSubmitUseCase } from '@attendance/domain/useCases/fixDailyRequest/ISubmitUseCase';
import { IUseCase as ILegalAgreementFetchOvertimeUseCase } from '@attendance/domain/useCases/legalAgreement/IFetchOvertimeUseCase';
import { IUseCase as ILegalAgreementRequestApprovalRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelApprovalUseCase';
import { IUseCase as ILegalAgreementRequestCancelRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelRequestUseCase';
import { IUseCase as ILegalAgreementRequestFetchListUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IFetchListUseCase';
import { IUseCase as ILegalAgreementRequestReapplyUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IReapplyUseCase';
import { IUseCase as ILegalAgreementRequestRemoveUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IRemoveUseCase';
import { IUseCase as ILegalAgreementRequestSubmitUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ISubmitUseCase';
import { IUseCase as IObjectivelyEventLogCreateUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ICreateUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyDeviationReasonsUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyDeviationReasonsUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyUseCase';
import { IUseCase as IObjectivelyEventLogFetchUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchUseCase';
import { IUseCase as IObjectivelyEventLogRemoveUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IRemoveUseCase';
import { IUseCase as IObjectivelyEventLogSaveDailyDeviationReasonUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISaveDailyDeviationReasonUseCase';
import { IUseCase as IObjectivelyEventLogSetToBeAppliedToDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISetToBeAppliedToDailyUseCase';
import { IUseCase as IStampTimePostUseCase } from '@attendance/domain/useCases/stampTime/IPostUseCase';
import { IUseCase as IDisplayFieldLayoutFetchTableUseCase } from '@attendance/domain/useCases/timesheet/IFetchTableUseCase';
import { IUseCase as ITimesheetFetchUseCase } from '@attendance/domain/useCases/timesheet/IFetchUseCase';

export type UseCases = {
  fetchTimesheet: ITimesheetFetchUseCase;
  reloadTimesheetOnly: ITimesheetFetchUseCase;
  fetchObjectivelyEventLogs: IObjectivelyEventLogFetchUseCase;
  fetchDailyObjectivelyEventLogDeviationReasons: IObjectivelyEventLogFetchDailyDeviationReasonsUseCase;
  setToBeAppliedToDailyObjectivelyEventLog: IObjectivelyEventLogSetToBeAppliedToDailyUseCase;
  reloadDailyObjectivelyEventLogs: IObjectivelyEventLogFetchDailyUseCase;
  reloadOneDailyObjectivelyEventLog: IObjectivelyEventLogFetchDailyUseCase;
  createObjectivelyEventLog: IObjectivelyEventLogCreateUseCase;
  removeObjectivelyEventLog: IObjectivelyEventLogRemoveUseCase;
  submitFixDailyRequest: IFixDailyRequestSubmitUseCase;
  submitFixDailyRequestWithClockOut: ISubmitFixDailyRequestWithClockOutUseCase;
  submitFixDailyRequestAndSaveDailyRecord: ISubmitFixDailyRequestAndSaveDailyRecordUseCase;
  cancelSubmittedFixDailyRequest: IFixDailyRequestCancelSubmittedUseCase;
  cancelApprovalFixDailyRequest: IFixDailyRequestCancelApprovalUseCase;
  saveDailyRecord: IDailyRecordSaveUseCase;
  saveDailyObjectivelyEventLogDeviationReason: IObjectivelyEventLogSaveDailyDeviationReasonUseCase;
  stampTime: IStampTimePostUseCase;
  fetchRestTimeReasons: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchListLegalAgreementRequest: ILegalAgreementRequestFetchListUseCase;
  fetchOvertimeLegalAgreement: ILegalAgreementFetchOvertimeUseCase;
  submitLegalAgreementRequest: ILegalAgreementRequestSubmitUseCase;
  cancelRequestLegalAgreementRequest: ILegalAgreementRequestCancelRequestUseCase;
  cancelApprovalLegalAgreementRequest: ILegalAgreementRequestApprovalRequestUseCase;
  removeLegalAgreementRequest: ILegalAgreementRequestRemoveUseCase;
  reapplyLegalAgreementRequest: ILegalAgreementRequestReapplyUseCase;
  fetchDailyFieldLayoutTable: IDisplayFieldLayoutFetchTableUseCase;
  saveFields: IDailyRecordSaveFieldsUseCase;
};

const service = UseCaseCollection<UseCases>("Timesheet-PC's UseCase Service");

export default service;
