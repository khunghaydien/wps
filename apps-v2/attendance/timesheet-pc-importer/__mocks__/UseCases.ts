import { UseCases } from '../UseCases';

export const service: UseCases = {
  fetchUserSetting: jest.fn(),
  fetchRestTimeReasons: jest.fn(),
  fetchRestTimeReasonsForBulk: jest.fn(),
  fetchLeaves: jest.fn(),
  fetchEarlyLeaveReasons: jest.fn(),
  fetchLateArrivalReasons: jest.fn(),
  fetchContractedWorkTimes: jest.fn(),
  saveTimesheet: jest.fn(),
  checkTimesheet: jest.fn(),
};

export default (): UseCases => service;
