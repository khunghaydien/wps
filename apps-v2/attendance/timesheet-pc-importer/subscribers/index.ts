import { Store } from 'redux';

import Events from '../Events';
import createBehaviors from './behaviors';
import UseCases from '@attendance/timesheet-pc-importer/UseCases';

export default (store: Store): void => {
  const behaviors = createBehaviors(store);

  // 社員切替後
  Events.selectedEmployee.subscribe(behaviors.cleanTimesheetRecords);

  // 所定情報取得後
  UseCases().fetchContractedWorkTimes.subscribe(behaviors.fetchRestTimesBulk);
};
