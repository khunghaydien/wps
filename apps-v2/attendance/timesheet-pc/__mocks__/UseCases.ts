import { UseCases } from '../UseCases';

export const methods: UseCases = {
  fetchTimesheet: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  reloadTimesheetOnly: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  fetchObjectivelyEventLogs: jest.fn().mockResolvedValue({
    objectivelyEventLogs: 'objectivelyEventLogs',
  }),
  setToBeAppliedToDailyObjectivelyEventLog: jest
    .fn()
    .mockResolvedValue('result'),
  reloadDailyObjectivelyEventLogs: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  reloadOneDailyObjectivelyEventLog: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  createObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  removeObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  stampTime: jest.fn().mockResolvedValue('result'),
  saveDailyRecord: jest.fn().mockResolvedValue('result'),
  saveDailyObjectivelyEventLogDeviationReason: jest
    .fn()
    .mockResolvedValue('result'),
  submitFixDailyRequest: jest.fn().mockResolvedValue('result'),
  submitFixDailyRequestWithClockOut: jest.fn().mockRejectedValue('result'),
  submitFixDailyRequestAndSaveDailyRecord: jest
    .fn()
    .mockRejectedValue('result'),
  cancelSubmittedFixDailyRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalFixDailyRequest: jest.fn().mockResolvedValue('result'),
  fetchRestTimeReasons: jest.fn().mockResolvedValue('result'),
  fetchDailyObjectivelyEventLogDeviationReasons: jest
    .fn()
    .mockResolvedValue('result'),
  fetchListLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchOvertimeLegalAgreement: jest.fn().mockResolvedValue('result'),
  submitLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  reapplyLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelRequestLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  removeLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchDailyFieldLayoutTable: jest.fn().mockResolvedValue('result'),
  saveFields: jest.fn().mockResolvedValue('result'),
};

export default (): UseCases => methods;
