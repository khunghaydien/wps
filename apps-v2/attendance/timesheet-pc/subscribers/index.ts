import { Store } from 'redux';

import Events from '../events';
import createBehaviors from './behaviors';
import UseCases from '@attendance/timesheet-pc/UseCases';

export default (store: Store): void => {
  const behaviors = createBehaviors(store);

  // 初期化後
  let initialized = false;
  Events.initialized.subscribe(({ fetchedTimesheet }) => {
    // 工数は UserSetting を見ているのでこの時に初期化するしかない
    behaviors.reloadTimeTrack(fetchedTimesheet);
    if (initialized) {
      return;
    }
    // 初期化が終わった後は UserSetting を常に持っているので reloadTimesheet で読み込んで問題ない
    UseCases().fetchTimesheet.subscribe(behaviors.reloadTimeTrack);
    initialized = true;
  });

  // 勤怠明細が更新された後
  Events.updatedDailyRecord.subscribe(() => {
    behaviors.reloadTimesheet();
    behaviors.reloadStampTime();
  });

  // 打刻された後
  Events.stampedTime.subscribe(behaviors.resetTimesheet);

  // 勤務表読み込み後
  UseCases().fetchTimesheet.subscribe(
    behaviors.reloadDailyObjectivelyEventLogs
  );

  // 勤務表読み込み後
  UseCases().fetchTimesheet.subscribe(behaviors.reloadDailyAllowanceRecords);

  // 勤務表読み込み後
  UseCases().fetchTimesheet.subscribe(
    behaviors.reloadListLegalAgreementRequest
  );

  // 勤務表読み込み後
  UseCases().fetchTimesheet.subscribe(behaviors.fetchDailyFieldLayoutTable);
};
