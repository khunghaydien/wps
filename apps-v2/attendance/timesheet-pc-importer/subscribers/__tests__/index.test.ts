import configureMockStore from 'redux-mock-store';

import snapshotDiff from 'snapshot-diff';

import Events from '../../Events';
import UseCases from '../../UseCases';
import subscribers from '..';
import EventEmitter from '@attendance/libraries/Event/emitter';

const createStore = configureMockStore();

jest.mock('../../UseCases', () => {
  const act = jest.requireActual('../../UseCases');
  const mocks = jest.requireActual('../../__mocks__/UseCases');
  act.default.register(mocks.service);
  return {
    __esModule: true,
    default: act.default,
  };
});

jest.mock('../behaviors', () => {
  const act = jest.requireActual('../behaviors');
  return {
    __esModule: true,
    default: jest.fn((store) => {
      const methods = act.methods;
      return Object.keys(methods).reduce((obj, key) => {
        obj[key] = (...args) => {
          store.dispatch({
            type: `METHOD: ${key}`,
            arguments: args,
          });
        };
        return obj;
      }, {});
    }),
  };
});

describe('Event:', () => {
  describe.each(Object.keys(Events))('%s:', (key) => {
    let before;
    let after;
    it('should subscribe', async () => {
      // Arrange
      const store = createStore({});
      await subscribers(store);

      // Act
      Events[key].publish();

      // Assert
      before = store.getActions();
      expect(before).toMatchSnapshot();

      EventEmitter.clear();
    });

    it('is shown diff', () => {
      expect(snapshotDiff(before, after)).toMatchSnapshot();
    });
  });
});

describe('UseCases:', () => {
  describe.each(Object.keys(UseCases()))('%s:', (key) => {
    let before;
    let after;
    it('should subscribe', async () => {
      // Arrange
      const store = createStore({});
      await subscribers(store);

      // Act
      EventEmitter.publish(UseCases()[key].eventName, {});

      // Assert
      before = store.getActions();
      expect(before).toMatchSnapshot();

      EventEmitter.clear();
    });

    it('is shown diff', () => {
      expect(snapshotDiff(before, after)).toMatchSnapshot();
    });
  });
});
