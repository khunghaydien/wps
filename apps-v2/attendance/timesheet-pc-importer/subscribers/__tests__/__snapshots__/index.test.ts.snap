// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event: selectedEmployee: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`Event: selectedEmployee: should subscribe 1`] = `
Array [
  Object {
    "arguments": Array [
      undefined,
    ],
    "type": "METHOD: cleanTimesheetRecords",
  },
]
`;

exports[`UseCases: checkTimesheet: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: checkTimesheet: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchContractedWorkTimes: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchContractedWorkTimes: should subscribe 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {},
    ],
    "type": "METHOD: fetchRestTimesBulk",
  },
]
`;

exports[`UseCases: fetchEarlyLeaveReasons: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchEarlyLeaveReasons: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchLateArrivalReasons: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchLateArrivalReasons: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchLeaves: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchLeaves: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchRestTimeReasons: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchRestTimeReasons: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchRestTimeReasonsForBulk: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchRestTimeReasonsForBulk: should subscribe 1`] = `Array []`;

exports[`UseCases: fetchUserSetting: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: fetchUserSetting: should subscribe 1`] = `Array []`;

exports[`UseCases: saveTimesheet: is shown diff 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected array but received undefined."
`;

exports[`UseCases: saveTimesheet: should subscribe 1`] = `Array []`;
